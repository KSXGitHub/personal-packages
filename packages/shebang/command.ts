export function command (shebang: string): string | null {
  if (!shebang.startsWith('#!')) return null
  shebang = shebang.slice(2).trim()
  const segments = shebang.split(/\s+/)
  while (segments[0].endsWith('/env') && segments.length > 1) {
    segments.shift()
  }
  return segments[0].split('/').slice(-1)[0]
}

export default command
