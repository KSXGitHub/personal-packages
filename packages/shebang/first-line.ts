import { createReadStream } from 'fs'
import { createInterface } from 'readline'

const READ_STREAM_OPTIONS = { encoding: 'utf8' } as const

export function firstLine(filename: string) {
  const stream = createReadStream(filename, READ_STREAM_OPTIONS)
  const reader = createInterface(stream)
  let result: string
  reader.once('line', line => {
    result = line
    reader.close()
  })
  return new Promise<string>(resolve => {
    reader.once('close', () => resolve(result))
  })
}

export default firstLine
