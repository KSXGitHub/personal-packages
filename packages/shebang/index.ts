export * from './command'
export * from './first-line'
export * from './shebang-command'
