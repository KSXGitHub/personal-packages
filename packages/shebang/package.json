{
  "name": "@khai96x/shebang",
  "version": "0.0.0",
  "description": "Extract shebang information from a file",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/KSXGitHub/personal-packages#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/KSXGitHub/personal-packages.git"
  },
  "bugs": {
    "url": "https://github.com/KSXGitHub/personal-packages/issues"
  },
  "keywords": [
    "shebang"
  ],
  "sideEffects": false,
  "dependencies": {
    "tslib": "^1.11.1",
    "@types/node": "^13.7.7"
  }
}
