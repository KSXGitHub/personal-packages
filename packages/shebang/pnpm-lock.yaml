lockfileVersion: 5.3

specifiers:
  '@types/node': ^16.0.0
  tslib: ^2.0.1

dependencies:
  '@types/node': 16.11.38
  tslib: 2.0.1

packages:

  /@types/node/16.11.38:
    resolution: {integrity: sha512-hjO/0K140An3GWDw2HJfq7gko3wWeznbjXgg+rzPdVzhe198hp4x2i1dgveAOEiFKd8sOilAxzoSJiVv5P/CUg==}
    dev: false

  /tslib/2.0.1:
    resolution: {integrity: sha512-SgIkNheinmEBgx1IUNirK0TUD4X9yjjBRTqqjggWCU3pUEqIk3/Uwl3yRixYKT6WjQuGiwDv4NomL3wqRCj+CQ==}
    dev: false
