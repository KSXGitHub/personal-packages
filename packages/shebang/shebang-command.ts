import command from './command'
import firstLine from './first-line'
export const shebangCommand = async (filename: string) => command(await firstLine(filename))
export default shebangCommand
