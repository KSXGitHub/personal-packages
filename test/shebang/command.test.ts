import { command } from '@khai96x/shebang'

it('matches snapshot', () => {
  const result = [
    '#!/bin/sh',
    '#!/usr/bin/sh',
    '#!/bin/env sh',
    '#!/usr/bin/env sh -',
    '#! /bin/bash',
    '#! /usr/bin/bash',
    '#! /bin/env bash',
    '#! /usr/bin/env bash -',
    'not-shebang',
  ].map(shebang => ({
    shebang,
    command: command(shebang),
  }))
  expect(result).toMatchSnapshot()
})
