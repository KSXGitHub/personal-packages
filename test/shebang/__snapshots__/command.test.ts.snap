// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  Object {
    "command": "sh",
    "shebang": "#!/bin/sh",
  },
  Object {
    "command": "sh",
    "shebang": "#!/usr/bin/sh",
  },
  Object {
    "command": "sh",
    "shebang": "#!/bin/env sh",
  },
  Object {
    "command": "sh",
    "shebang": "#!/usr/bin/env sh -",
  },
  Object {
    "command": "bash",
    "shebang": "#! /bin/bash",
  },
  Object {
    "command": "bash",
    "shebang": "#! /usr/bin/bash",
  },
  Object {
    "command": "bash",
    "shebang": "#! /bin/env bash",
  },
  Object {
    "command": "bash",
    "shebang": "#! /usr/bin/env bash -",
  },
  Object {
    "command": null,
    "shebang": "not-shebang",
  },
]
`;
